_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,140 @@
+---
+layout: singlepage-overview
+title: Binary Compatibility for library authors
+
+discourse: true
+permalink: /tutorials/:title.html
+---
+
+## Introduction
+
+A diverse and comprehensive set of libraries is important to any productive software ecosystem. While it is easy to develop and distribute Scala libraries, good library authorship goes far
+beyond just writing code and publishing them.
+
+In this guide, we will cover the important topic of **Binary Compatibility**:
+
+* How binary incompatibility can cause production failures in your applications
+* How library authors can avoid breaking binary compatibility, and/or convey breakages clearly to library users when they happen
+
+Before we start, first we need to understand how code is compiled and executed on the Java Virtual Machine (JVM).
+
+## The JVM execution model
+
+Code compiled to run on the JVM is compiled to a platform-independent format called **JVM bytecode** and stored in **Class File** format (with `.class` extension) and these class files are stored
+in JAR files. The bytecode is what we refer to as the **Binary** format.
+
+When application or library code is compiled, their bytecode invokes named references of classes/methods from their dependencies instead of including the dependencies' actual bytecode
+(unless inlining is explicitly requested). During runtime, the JVM classloader will search through the provided class files for classes/methods referenced by and invoke them.
+
+Let's illustrate with an example:  
+
+We got an application `App` that depends on `A` which itself depends on library `C`. When starting the application we need to provide the class files
+for all of `App`, `A` and `C` (something like `java -cp App.jar:A.jar:C.jar:. MainClass`). If we did not provide `C.jar`, or if we provided a `C.jar` that does not contain certain classes/methods
+which `A` expected to exist in library `C`, we will get an exception when our code attempt to invoke the missing classes/methods.
+
+This is what we call **Binary Incompatibility Errors** - The bytecode interface used for compilation differs and is incompatible with the bytecode provided during runtime.
+
+## What are Evictions, Source Compatibility and Binary Compatibility?
+
+Since the classloader only loads the first match of a class, having multiple versions of the same library in the classpath is redundant. 
```

Comment:
What about "When the JVM loads a given class, the first class file that is matched in the classpath is used. If another library of the classpath also has a class file with the same name, it'll be ignored. It is good practice to make sure that this case doesn't happen, since it can only cause trouble."?

Reply:
Yeap I think your description is much more accurate. I'd not use "it is good practice" though as most of the time the tooling will take care of it for you by eviction.